vllm/engine/llm_engine.py

```diff
@@ -1376,11 +1372,12 @@ def _process_model_outputs(self,
             request_output = RequestOutputFactory.create(seq_group)
             ctx.request_outputs.append(request_output)
 
-        # For async + multi-step, do stats only on the last output.
-        # Otherwise, do stats if the execution is async
-        do_stats = is_multi_step or is_async
+        # Immediately process request outputs here (if callback is given)
+        if len(ctx.request_outputs
+               ) > 0 and self.process_request_outputs_callback is not None:
+            self.process_request_outputs_callback(ctx.request_outputs)
 
-        if do_stats:
+        if is_async:
```

Comment:
Why only record metrics when async?

Reply:
In the non-async, it is explicitly recorded in the llm_engine or async_llm_engine, here it is only for async case

Reply:
Thanks for the clarification. Would you mind adding a comment to avoid future confusion?

Reply:
Added